### ISSUE TITLE
[Gemma4/Wormhole] Deterministic garbled long-prose output at specific encoded lengths >=100k on T3K (12B & 31B); BH clean at same lengths

### ISSUE THREAD

**Issue by arginugaTT** [tenstorrent/tt-metal#55102]
## Summary
Gemma4 long-prose generation on **Wormhole T3K** produces deterministic garbled/degenerate output at specific encoded prompt lengths ≥ ~100k tokens. **WH-local**: Blackhole (emulated-LB Galaxy, 8-point walk 100,462–130,816 including 106,500 / 108,000 / 109,857) is clean at the same lengths. Both 31B and 12B affected at different points.

## Point map (branch `arg/gemma4_coherence_fix` @ c5e8d23845f, T3K, Frankenstein quote task, greedy, x2 byte-equality per point)
| model | encoded | verdict |
|---|---|---|
| 31B | 91,052 / 97,097 / 100,462 / 104,954 | clean, byte-equal |
| 31B | **100,793** | GARBLED (`'s la's` loop), deterministic |
| 31B | **109,857** | GARBLED (`I had't been...`), deterministic |
| 31B | **127,574** | GARBLED (`lapped` flavor) |
| 12B | 12,288 / 16,384 / ~106k | clean |
| 12B | ~120k | coherent but rep-divergent |
| 12B | **127,574** | fully degenerate token salad |

Note the **non-monotonicity** (100,462 clean → 100,793 corrupt → 104,954 clean → 109,857 corrupt): there is no single length threshold.

## Evidence
- **Path-independent**: metal demo and vLLM serving produce *byte-identical* garble at 109,857 → in the shared model forward (prefill), not serving state.
- **Chunk-invariant**: `GEMMA4_PREFILL_CHUNK_SIZE` 8192→4096 at 109,857 → byte-identical output. Not per-Q-chunk config ("absolute-position family").
- **Deterministic** per exact input (byte-equal across reps and across metal/server).
- **Build-stable**: 100,793 produces the identical `'s la's` loop on the pre-coherence-fix build.
- **Cross-platform control**: BH clean at these lengths (shared ledger §10).
- Ruled out: block-table width boundaries (98,304 = 1536 blocks crossed clean), 64k page-table width, prefill chunk count, sliding-stride multiples (98,304 = 6x16,384 clean).

## Suspect class
WH-conditional seq-len-dependent kernel/program config in the prefill path (precedent: `PREFILL_SDPA_HARD_MAX = 32768` — "non-chunked SDPA silently wrong past 2^15"). Static audit of gemma4 attention configs found no WH-conditional constant in the 100–130k range yet; the defect may live below the model layer (SDPA/matmul program-config selection at large cache_len on WH).

## Repro
`env -u CI MESH_DEVICE=T3K HF_MODEL=google/gemma-4-31B-it pytest models/demos/gemma4/demo/text_demo_v2.py -k long-context-128k` with the standard 128k prompt json (encodes to 100,793) → item B of the quote answer is a deterministic `'s la's` loop.

**Comment by arginugaTT** [comment]
## R2 bisect + audit results (update)

**Onset bisected to <=11 tokens: (100,462 -> 100,473] encoded** (31B/T3K, quote task, greedy, deterministic). Bisection points (chars of pg84.txt prefix -> encoded -> verdict): 437,010 -> 100,462 clean; 437,239 -> 100,473 GARBLED ("I had'd a laudanum"); 437,467 -> 100,513 GARBLED ("had't been a lapped"); 437,925 -> 100,608 GARBLED ('s la's loop); full book -> 100,793 GARBLED ('s la's loop).

Three mechanism-relevant properties:
1. **Severity grades continuously with length** (one-token grammar slip -> mixed misquote -> degenerate loop) — not a hard cliff.
2. **The 11-token onset window contains no multiple of 32 or 64** (100,448 and 100,480 both outside) — not aligned to tile/block geometry.
3. Combined with earlier findings — chunk-size invariance (Q-chunk 8192 vs 4096 byte-identical) and non-monotonic task visibility (104,954 verbatim-clean above the onset) — this fits **progressive numerical degradation at large K crossing task-visibility threshold**, not a program-config selection bug.

**Fidelity audit: model layer exonerated.** `models/demos/gemma4/tt/compute_config.py` sets SDPA HiFi4 + fp32_dest_acc unconditionally, `exp_approx_mode=False`; no arch conditionals anywhere in the gemma4 attention path. Since BH is verbatim-clean at 106,500/108,000/109,857 with identical model code and settings, the divergence is **beneath the model layer**: arch-specific SDPA kernel numerics at large K on Wormhole (candidate: fp32_dest_acc handling differences WH vs BH). Suggested next probe for kernel owners: PCC of chunked-SDPA output vs fp32 reference as a function of K length on WH, same sweep on BH — prediction: WH degrades smoothly past ~100k, BH does not.

Repro note: `e7e3c8771aa` (repeat_context honored) does not affect the repro — the standard 128k json has no repeat_context key.

**Comment by arginugaTT** [comment]
## W1: measured PCC-vs-K curve (WH T3K, shipped chunked-SDPA config)

Op-level probe: `chunked_prefill_sdpa` exactly as production (HiFi4 + fp32_dest_acc, grid 8x4, q/k_chunk 128, paged KV, Q pre-scaled / scale=1.0), bf16 randn inputs, PCC of the **last 128 Q rows** (the rows feeding last-token logits) vs fp32 host reference.

| K | PCC (tail-128) | max_abs_err |
|---|---|---|
| 32,768 | 0.999660 | 0.0017 |
| 65,536 | 0.999404 | 0.0026 |
| 98,304 | 0.999142 | 0.0033 |
| 100,352 | 0.999141 | 0.0032 |
| 104,448 | 0.999124 | 0.0031 |
| 110,080 | 0.999001 | 0.0030 |
| 131,072 | 0.998719 | 0.0046 |

Reading:
- **Smooth, monotonic degradation with K — no cliff, no knee at ~100k.** Consistent with the graded end-to-end symptom; refines the verdict: the ~100k onset is where ~48 layers of compounding crosses task visibility, not where the op changes behavior.
- Falsifiability: the BH mirror of this exact sweep is being run. If BH's curve matches WH's, this op is NOT the WH/BH differentiator and the kernel-numerics verdict falls back under review.
- Caveats: synthetic randn distribution; single-layer PCC ~0.9987 at 131k is individually high — the claim rests on compounding, which an end-to-end (model-level) fp32-reference comparison would nail down if the cross-platform delta confirms.
- Probe note for reproducers: the op rejects non-1.0 `scale` on this build (empty-message throw; production pre-scales Q upstream and passes scale=1.0 - see prefill.py "SDPA (causal prefill, scale=1.0)").

**Comment by arginugaTT** [comment]
## W1 extended: 28-point PCC-vs-K curve (WH T3K) — degradation is super-linear in K

Extends the earlier 7-point curve (comment above): both gemma4 SDPA shapes (head_dim 512 global / 256 sliding-causal), two seeds each, K through 262,144. Same probe: shipped `chunked_prefill_sdpa` config (HiFi4 + fp32_dest_acc, scale=1.0 with pre-scaled Q), bf16 randn, PCC of the tail-128 rows vs fp32 reference.

| K | PCC hd512 (s0/s1) | PCC hd256 (s0/s1) | maxerr hd512 |
|---|---|---|---|
| 32,768 | 0.999660 / 0.999657 | 0.999658 / 0.999616 | 0.0017 |
| 65,536 | 0.999457 / 0.999410 | 0.999430 / 0.999424 | 0.0025 |
| 98,304 | 0.999159 / 0.999142 | 0.999160 / 0.999085 | 0.0033 |
| 131,072 | 0.998919 / 0.998362 | 0.998803 / 0.998453 | 0.0039 |
| 163,840 | 0.997181 / 0.997815 | 0.996774 / 0.996705 | 0.0069 |
| 200,704 | 0.994239 / 0.995083 | 0.993473 / 0.993631 | 0.0075 |
| **262,144** | **0.990063 / 0.989988** | **0.989559 / 0.990012** | **0.0107** |

Reading:
- **Super-linear decay**: the loss rate roughly doubles per +65k K past 131k (Δ(1-PCC): ~0.0009 to 131k, then ~0.0016, ~0.0035, ~0.0047 per step). Not gentle drift — a compounding accumulation effect in the K-loop.
- **head_dim-independent**: 512 and 256 curves overlap within seed noise → mechanism lives in the K-length accumulation (softmax running-max/denominator or PV accumulation), not in the head-width math.
- **Seed-stable** → systematic, not noise.
- Consistency check with the end-to-end symptom: single-layer PCC ~0.9990 at the ~100k garble onset compounds across 48 layers to visible token flips; by 262k a single layer is at 0.990.
- The BH mirror of this sweep remains the cross-platform falsifier; if BH tracks this curve, the WH-only garble needs a different differentiator (but the end-to-end evidence — BH verbatim-clean at 131k/262k prose where WH garbles — still stands either way).
